### ISSUE TITLE
OpenJ9 (Linux) JVM opens more files than RI

### ISSUE THREAD

**Issue by JasonFengJ9** [eclipse/openj9#4084]
Discovered during investigation of https://github.com/eclipse/openj9/issues/3970.

The test code within that issue produces following results for OpenJ9:
<pre>
openjdk version "1.8.0_192"
OpenJDK Runtime Environment (build 1.8.0_192-201812170453-b12)
Eclipse OpenJ9 VM (build master-f9ab45bfe, JRE 1.8.0 Linux amd64-64-Bit Compressed References 20181217_142 (JIT enabled, AOT enabled)
OpenJ9   - f9ab45bfe
OMR      - 0d514697
JCL      - 7105b0a753 based on jdk8u192-b12)

ofdc : <b>12</b> / Java Version 1.8.0_192 Eclipse OpenJ9 / OS Linux 3.10.0-693.1.1.el7.x86_64
</pre>
While RI produces:
<pre>
java version "1.8.0_181"
Java(TM) SE Runtime Environment (build 1.8.0_181-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.181-b13, mixed mode)

ofdc : <b>5</b> / Java Version 1.8.0_181 Oracle Corporation / OS Linux 3.10.0-693.1.1.el7.x86_64
</pre>

Created another testcase which simply performs a looping:
```
		for (;;) {
			Thread.sleep(3000);
		}
```
`/proc/$(process_id)/fd` shows the files opened by JVM:
<pre>
<b>OpenJ9 JVM</b>
lrwx------ 1 Dec 17 16:33 0 -> /dev/pts/0
lrwx------ 1 Dec 17 16:33 1 -> /dev/pts/0
lr-x------ 1  Dec 17 16:33 10 -> /PathTo/TestHelloLoop.jar
lrwx------ 1 Dec 17 16:33 2 -> /dev/pts/0
lr-x------ 1  Dec 17 16:33 3 -> /proc/1/cgroup
lr-x------ 1  Dec 17 16:33 4 -> /proc/1/cgroup
lr-x------ 1  Dec 17 16:33 5 -> /jdk8u192-b12/jre/lib/rt.jar
lr-x------ 1  Dec 17 16:33 6 -> /jdk8u192-b12/jre/lib/resources.jar
lr-x------ 1  Dec 17 16:33 7 -> /jdk8u192-b12/jre/lib/jsse.jar
lr-x------ 1  Dec 17 16:33 8 -> /jdk8u192-b12/jre/lib/charsets.jar
lr-x------ 1  Dec 17 16:33 9 -> /jdk8u192-b12/jre/lib/jce.jar

<b>RI JVM</b>
lrwx------ 1  Dec 17 16:28 0 -> /dev/pts/0
lrwx------ 1  Dec 17 16:28 1 -> /dev/pts/0
lrwx------ 1  Dec 17 16:28 2 -> /dev/pts/0
lr-x------ 1  Dec 17 16:28 3 -> /jdk1.8.0_181/jre/lib/rt.jar
lr-x------ 1  Dec 17 16:28 4 -> /PathTo/TestHelloLoop.jar
</pre>

Note: `OpenJ9` (with https://github.com/eclipse/omr/pull/3363) has same results as `RI`.
<pre>
<b>OpenJ9 JVM</b>
ofdc : <b>5</b> / Java Version 11.0.1-internal <b>Eclipse OpenJ9</b> / OS Mac OS X 10.14.1

<b>RI JVM</b>
ofdc : <b>5</b> / Java Version 11 <b>Oracle</b> Corporation / OS Mac OS X 10.14.1
</pre>

**Comment by pdbain-ibm** [comment]
Running liberty* with OpenJDK needs 275-280 simultaneous file handles (determined via `ulimit -n <limit>` and parsing the output of `strace -f -e trace=file -e trace=desc`.  OpenJ9 needs ~290.
Note that OpenJDK (but not OpenJ9) overrides the soft limit:
```
26865 getrlimit(RLIMIT_NOFILE, {rlim_cur=200, rlim_max=1024*1024}) = 0
26865 setrlimit(RLIMIT_NOFILE, {rlim_cur=1024*1024, rlim_max=1024*1024}) = 0
```
so `ulimit -Sn` has no effect.

At steady state, both processes had roughly the same number of files (~148) open. The majority of the files are liberty files, sockets, and pipes.  

\* The details:
Run Liberty like so:
```
git clone https://github.com/kwsutter/dukes-microprofile
cd dukes-microprofile
mvn clean install
cd dukes-liberty
mvn liberty:run-server
```
Java versions are:
- OpenJ9:
openjdk version "11.0.2" 2018-10-16
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.2+6-201901221910)
Eclipse OpenJ9 VM AdoptOpenJDK (build master-b5e3c2a4c, JRE 11 Linux amd64-64-Bit Compressed References 20190122_103 (JIT enabled, AOT enabled)
OpenJ9   - b5e3c2a4c
OMR      - 8b8cbdaf
JCL      - 007fb48620 based on jdk-11.0.2+6)

- OpenJDK:
openjdk version "11.0.1" 2018-10-16
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.1+13)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.1+13, mixed mode)

**Comment by ashu-mehra** [comment]
These two:
```
lr-x------ 1  Dec 17 16:33 3 -> /proc/1/cgroup
lr-x------ 1  Dec 17 16:33 4 -> /proc/1/cgroup
```
should go away after https://github.com/eclipse/omr/pull/3497 is merged.

**Comment by kwsutter** [comment]
FYI...  Setting a larger Soft limit actually helped with the sample that @pdbain-ibm posted above.  At least in my environment, it did...  By default, my macbook pro has a Soft limit of 256.  But, when I set the limit to 1024 via ```ulimit -Sn 1024```, the example started to work just fine.  All of that was with OpenJ9.  If I used Hotspot, I had no issues even when the Soft limit was set at the default of 256.

**Comment by pdbain-ibm** [comment]
Now OpenJ9 opens _fewer_ files.
Tested with:
_openjdk version "11.0.2-internal" 2018-10-16
OpenJDK Runtime Environment (build 11.0.2-internal+0-adhoc.pdbain.openj9-openjdk-jdk11)
Eclipse OpenJ9 VM (build cleanup-ab9bc1883, JRE 11 Mac OS X amd64-64-Bit Compressed References 20190124_000000 (JIT enabled, AOT enabled)
OpenJ9   - ab9bc1883
OMR      - 7e6db886
JCL      - 007fb48620 based on jdk-11.0.2+7)_
`ofdc : 4 / Java Version 11.0.2-internal Eclipse OpenJ9 / OS Mac OS X 10.13.6`

with
_openjdk version "11" 2018-09-25
OpenJDK Runtime Environment AdoptOpenJDK (build 11+28)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11+28, mixed mode)_
`ofdc : 5 / Java Version 11 Oracle Corporation / OS Mac OS X 10.13.6`

**Comment by aguibert** [comment]
FYI I've seen 2 other people hit this issue on MacOS

**Comment by aguibert** [comment]
WORKAROUND: Set the open file limit to a higher value, such as 1024:

```
$ ulimit -Sn 1024
```

**Comment by evelinec** [comment]
You can use the Open Liberty guides, ie, https://openliberty.io/guides/microprofile-jwt.html to reproduce this problem on MacOS with AdoptedOpenJDK11 OpenJ9 JVM. The `ulimit -Sn 1024` workaround fixes the problem. Is there any programmatical workaround we can use?

**Comment by DanHeidinga** [comment]
With the fix to set the soft limit to the hard limit in #4442, I'm going to close this as fixed

**Comment by aguibert** [comment]
since a lot of people are hitting this issue, I will post that it's fixed as of `openj9-0.13.0`
